internal/buffer/in_message.go

```diff
@@ -52,11 +53,56 @@ func NewInMessage() *InMessage {
 	}
 }
 
+func readAll(r io.Reader, dest []byte) (int, error) {
+	offset := 0
+
+	for offset < len(dest) {
+		// read the remaining buffer
+		n, err := r.Read(dest[offset:])
+		if n == 0 && err == nil {
+			// remote fd closed
+			return n, err
+		}
+		offset += n
+		if err != nil {
+			return offset, err
+		}
+	}
+	return offset, nil
+}
+
+var readLock sync.Mutex
+
+func (m *InMessage) ReadSingle(r io.Reader) (int, error) {
+	readLock.Lock()
+	defer readLock.Unlock()
+
+	// read request length
+	if _, err := readAll(r, m.storage[0:4]); err != nil {
+		return 0, err
+	}
+
+	l := m.Header().Len
+	// read remaining request
+	if n, err := readAll(r, m.storage[4:l]); err != nil {
+		return n, err
+	}
+	return int(l), nil
+}
+
 // Initialize with the data read by a single call to r.Read. The first call to
 // Consume will consume the bytes directly after the fusekernel.InHeader
 // struct.
 func (m *InMessage) Init(r io.Reader) error {
-	n, err := r.Read(m.storage[:])
+
+	var n int
+	var err error
+	if fusekernel.IsPlatformFuseT {
```

Comment:
Why the difference in behavior here? Can you explain what the difference in semantics are and why fuse-t differs?